File: ts/src/siggate.ts
/* Copyright (c) 2021-2026 Richard Rodger, MIT License */


import type { Val } from './type'
import { AontuContext } from './ctx'
import { makeNilErr } from './err'

import { funcSig, renderSig } from './sig'
import type { FuncSig, ArgSig } from './sig'

import {
  BigDecimal,
  BigInteger,
  Float,
  Integer,
  Path,
  kindSubsumes,
} from './val/ScalarKindVal'


const SIG_KIND = new Map<string, any>([
  ['string', String],
  ['number', Number],
  ['integer', Integer],
  ['float', Float],
  ['biginteger', BigInteger],
  ['bigdecimal', BigDecimal],
  ['boolean', Boolean],
  ['path', Path],
])


function gateWords(type: string): any[] | undefined {
  const out: any[] = []
  for (const word of type.split('|')) {
    const marker = SIG_KIND.get(word)
    if (undefined === marker) {
      return undefined
    }
    out.push(marker)
  }
  return out
}


// The declared type admits a driven Val when the Val is a concrete
// scalar whose leaf kind is, or sits below, one of the declared
// words -- the same walk subsumption makes, so `number` admits every
// numeric leaf and `string` admits a path value.
function admits(markers: any[], arg: any): boolean {
  const leaf: any = arg.superior?.()
  if (true !== arg.isScalar || true !== leaf?.isScalarKind) {
    return false
  }
  for (const marker of markers) {
    if (marker === leaf.peg || kindSubsumes(marker, leaf.peg)) {
      return true
    }
  }
  return false
}


// The gate. Answers the func_arg refusal, or undefined to let the
// call resolve.
function sigRefuse(
  ctx: AontuContext, fn: any, args: Val[]
): Val | undefined {
  const sig: FuncSig | undefined = funcSig[fn.funcname()]

  // key() reads its level off the written peg and `key_level` names
  // what is wrong with a bad one; the gate leaves the meaning where
  // it lives.
  if (undefined === sig || 'key' === sig.name) {
    return undefined
  }

  for (let i = 0; i < sig.args.length; i++) {
    const a: ArgSig = sig.args[i]
    if (true === a.rest) {
      break
    }
    if ('value' !== a.mode) {
      continue
    }
    const markers = gateWords(a.type)
    if (undefined === markers) {
      continue
    }
    const arg: any = args[i]
    if (undefined === arg || true === arg.isNil || true !== arg.done) {
      continue
    }
    const shaped = (true === arg.isScalar) ||
      (true === arg.isMap) || (true === arg.isList) ||
      (true === arg.isScalarKind)
    if (shaped && !admits(markers, arg)) {
      return makeNilErr(ctx, 'func_arg', fn, arg, undefined, {
        func: sig.name,
        sig: renderSig(sig),
        arg: a.name,
        argn: '' + (i + 1),
        got: arg.canon,
      })
    }
  }

  return undefined
} /* node:coverage ignore next 5 */


export {
  sigRefuse,
}

File: go/siggate.go

package aontu


import "strings"

// sigKindWords maps the scalar-kind words the gate enforces to their
// lattice kinds. A declared type is gate-checkable only when EVERY
// union word is here.
var sigKindWords = map[string]Kind{
	"string":     KindString,
	"number":     KindNumber,
	"integer":    KindInteger,
	"float":      KindFloat,
	"biginteger": KindBigInteger,
	"bigdecimal": KindBigDecimal,
	"boolean":    KindBoolean,
	"path":       KindPath,
}

func sigGateKinds(atype string) []Kind {
	words := strings.Split(atype, "|")
	out := make([]Kind, 0, len(words))
	for _, word := range words {
		k, ok := sigKindWords[word]
		if !ok {
			return nil
		}
		out = append(out, k)
	}
	return out
}

func sigAdmits(kinds []Kind, arg Val) bool {
	sv, ok := arg.(*ScalarVal)
	if !ok {
		return false
	}
	for _, k := range kinds {
		if k == sv.kind || kindSubsumes(k, sv.kind) {
			return true
		}
	}
	return false
}

// sigRefuse answers the func_arg refusal, or nil to let the call
// resolve. Mirrors sigRefuse in ts/src/siggate.ts.
func sigRefuse(ctx *Ctx, f *FuncVal, args []Val) Val {
	sig, ok := funcSig[f.name]

	// key() reads its level off the written peg and `key_level` names
	// what is wrong with a bad one; the gate leaves the meaning where
	// it lives.
	if !ok || "key" == sig.Name {
		return nil
	}

	for i, a := range sig.Args {
		if a.Rest {
			break
		}
		if ModeValue != a.Mode {
			continue
		}
		kinds := sigGateKinds(a.Type)
		if nil == kinds {
			continue
		}
		if len(args) <= i {
			continue
		}
		arg := args[i]
		if nil == arg || arg.Nil() || DONE != arg.Dc() {
			continue
		}
		shaped := false
		switch arg.(type) {
		case *ScalarVal, *MapVal, *ListVal, *ScalarKindVal:
			shaped = true
		}
		if shaped && !sigAdmits(kinds, arg) {
			out := makeNilErrFull(ctx, "func_arg", f, arg, "", map[string]string{
				"func": sig.Name,
				"sig":  renderSig(sig),
				"arg":  a.Name,
				"argn": itoa(i + 1),
				"got":  arg.Canon(),
			})
			return out
		}
	}

	return nil
}
